### PR TITLE
libnetwork/drivers/overlay: make VNI matcher IPv6-compatible

### PR DESCRIPTION
**Pull request by corhere**
**- What I did**
Made the VNI matcher BPF program compatible with ip6tables in addition to iptables.

**- How I did it**
Used Linux BPF extensions to locate the offset of the VXLAN header within the packet so that the same BPF program works with VXLAN packets received over either IPv4 or IPv6.

**- How to verify it**
A new automated test verifies the BPF program with the IPv4 stack. A manual end-to-end test (documented in a comment in the automated test's source) can be used to verify the program matches expected packets in a live ip6tables rule.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: libnetwork/drivers/overlay/bpf_test.go [hunk 1]
```diff
@@ -12,3 +24,218 @@ func FuzzVNIMatchBPFDoesNotPanic(f *testing.F) {
...
+	// libnetwork so this whole BPF program could be replaced with a native
+	// nftables '@th' match expression.
+	//
+	// The filter could be manually e2e-tested for both IPv4 and IPv6 by
+	// programming ip[6]tables rules which log matching packets and sending
+	// test packets loopback using netcat. All the necessary information
+	// (bytecode and acceptable test vector) is logged by this test.
+	//
+	//     $ sudo ip6tables -A INPUT -p udp -s ::1 -d ::1 -m bpf \
+	//         --bytecode '7,32 0 0 4294963252,7 0 0 0,64 0 0 4,116 0 0 8,21 1 0 4096,6 0 0 0,6 0 0 4294967295' \
```
Comment (neersighted) [inline libnetwork/drivers/overlay/bpf_test.go:51]: I'm not sure including the literal bytecode in the comment is a good idea, as future changes could cause it to become out of date. We could add a caveat here, or just substitute `[bytecode]` and some advice on how to use the Go assembler to output the bytecode for the above program.

Comment (corhere) [reply libnetwork/drivers/overlay/bpf_test.go]: Fair enough. I assumed a savvy developer who could recognize the implication that this was merely a worked example since the VNI is constant is baked into the program, but that might not be a safe assumption.

Comment (neersighted) [reply libnetwork/drivers/overlay/bpf_test.go]: I don't think that's necessary at all if it shows up in the verbose log.

**Review by akerouanton — CHANGES_REQUESTED** [review]
What's the reason for not testing IPv6 automatically?

File: libnetwork/drivers/overlay/bpf_linux_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+				pkt = append(pkt, []byte{0xde, 0xad, 0xbe, 0xef}...)
+				if _, err := sender.Write(pkt); err != nil {
+					t.Fatal(err)
+				}
+
+				rpkt, ok := readUDPPacketFromRawSocket(t, pc, saddr, daddr)
+				// Sanity check: the only packets readUDPPacketFromRawSocket
+				// should return are ones we sent.
+				if ok && !bytes.Equal(pkt, rpkt) {
+					t.Fatalf("received unexpected packet: % x", rpkt)
```
[the remark this answers is not in the document]
Comment (corhere) [reply libnetwork/drivers/overlay/bpf_linux_test.go:111]: That is not a typo

**Comment by corhere** [comment]
> What's the reason for not testing IPv6 automatically?

I documented the reason on lines 64-68 of bpf_linux_test.go

File: libnetwork/drivers/overlay/bpf_linux_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (akerouanton) [reply libnetwork/drivers/overlay/bpf_linux_test.go:111]: Ah, didn't knew %x supported a space in the middle. Thanks.

**Comment by akerouanton** [comment]
Let me rephrase my original comment: why did you choose to write a unit test instead of an e2e test, if the former doesn't allow testing the bpf rule with IPv6? Is there any particular reason for that?

**Comment by corhere** [comment]
We don't have an e2e test harness for moby and Swarm does not currently support IPv6 so I would have to boil the ocean before I could even begin to write an e2e test for the bpf rule.

**Review by tianon — APPROVED** [review]
Talked over this in the maintainers meeting and it looks solid to me, but I did wonder if the test works on the old implementation too, so @corhere has promised to test that and if it works, he'll swap the commits so the test gets added before the code change. :smile:

**Comment by corhere** [comment]
I have verified that the new test passes with the old implementation.

**Review by akerouanton — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
copying @akerouanton's LGTM 👍
